src/xvectorbin/nnet3-xvector-signal-perturb-egs.cc

```diff
@@ -73,16 +73,71 @@ void ApplyPerturbation(XvectorPerturbOptions opts,
   }
   // Perturb speed of signal egs
   Matrix<BaseFloat> warped_distorted_shifted_egs(rand_distort_shifted_egs);
-  if (opts.max_time_stretch != 0.0) 
-    perturb_xvector.TimeStretch(rand_distort_shifted_egs, 
-                                &warped_distorted_shifted_egs);
+  ////if (opts.max_time_stretch != 0.0) 
+  ////  perturb_xvector.TimeStretch(rand_distort_shifted_egs, 
+  ////                              &warped_distorted_shifted_egs);
    
   // If nagation is true, the sample values are randomly negated
   // with some probability.
-  if (opts.negation) {
+  ////if (opts.negation) {
    
+  ////}
+}
+
+// add
+// This function add the noise to the orginial signal. We should not normalize 
+// the signal level of the orginial signal. According to SNR, we rescale the noise
+// and add it. So that the perturbed signal is created. 
+void ApplyAddAdditiveNoise(const int32 &SNR,
+                           const Matrix<BaseFloat> &input_eg,
+                           const Matrix<BaseFloat> &noise_eg,
+                           Matrix<BaseFloat> *perturb_eg) {
+  // In the version, we ask the noise_cols >= input_cols. If mfcc, the cols are equal.
+  // If raw data, we ask the noise_cols > input_cols.
+  int32 input_rows = input_eg.NumRows(), input_cols = input_eg.NumCols();  
+  KALDI_ASSERT(noise_eg.NumCols() >= input_cols);
+
+  // According to the rows of noise_eg, form the noise_mat
+  // repeat the noise_eg blocks to have a new block which is longer than input_eg
+  Matrix<BaseFloat> noise_mat;
+  if (noise_eg.NumRows() < input_rows) {
+    int32 repeat_times = (input_rows / noise_eg.NumRows()) + 1;
+    noise_mat.Resize(noise_eg.NumRows() * repeat_times, noise_eg.NumCols());
+    for (int32 i = 0; i < repeat_times; ++i) {
+      noise_mat.Range(i*noise_eg.NumRows(), noise_eg.NumRows(), 
+                      0, noise_eg.NumCols()).CopyFromMat(noise_eg);
+    }
+  } else {
+    noise_mat.Resize(noise_eg.NumRows(), noise_eg.NumCols());
+    noise_mat.CopyFromMat(noise_eg);
   }
+
+  // select the noise range
+  int32 noise_rows = noise_mat.NumRows(), noise_cols = noise_mat.NumCols();
+  int32 start_row_ind = RandInt(0, noise_rows - input_rows),
+        start_col_ind = RandInt(0, noise_cols - input_cols); 
+  Matrix<BaseFloat> selected_noise_mat(input_rows, input_cols);
+  selected_noise_mat.AddMat(1.0, noise_mat.Range(start_row_ind, input_rows,
+                                                  start_col_ind, input_cols));
+  // compute the energy of noise and input
+  Matrix<BaseFloat> input_energy_mat(input_rows, input_cols);
+  input_energy_mat.AddMatMatElements(1.0, input_eg, input_eg, 1.0);
```

Comment:
It is not a good idea to design the code like this. You should write this function in signal-distort.h and add-noise and snr  should be added as options to XvectorPerturbationOptions struct.
The function should be ApplyAdditiveNoise(const VectorBase<BaseFloat> input, const VectorBase<BaseFloat> noise,  BaseFloat snr, Vector<BaseFloat> *noisy_input)

In class PerturbXvectorSignal, you have applyDistortion which is a general function, which applies all type of distortions to input.
Then it applies distortions w.r.t opts_.

You need to add a function PerturbExamples(const XvectorOptions opts, const Matrix<BaseFloat> &input_egs, Matrix<BaseFloat> *perturbed_egs)
and this function called in nnet3-xvector-signal-perturb.cc and it generates object from PerturbXvectorSignal and vectorize the input and calls ApplyDistortion to apply different type of distortions on input.